feat: Support for userCol and itemCol as string types in SAR model

dciborow opened the pull request:
Fixes #2275

Add support for `userCol` and `itemCol` as string types in the SAR model.

* **Python Files:**
  - Add `core/src/main/python/synapse/ml/recommendation/SAR.py` to handle string `userCol` and `itemCol`.
  - Modify `core/src/main/python/synapse/ml/recommendation/SARModel.py` to handle string `userCol` and `itemCol` in the `recommendForUserSubset` function.

* **Scala Files:**
  - Modify `core/src/main/scala/com/microsoft/azure/synapse/ml/recommendation/SAR.scala` to handle string `userCol` and `itemCol` in the `calculateUserItemAffinities` and `calculateItemItemSimilarity` functions.
  - Modify `core/src/main/scala/com/microsoft/azure/synapse/ml/recommendation/SARModel.scala` to handle string `userCol` and `itemCol`.

* **Tests:**
  - Update `core/src/test/python/synapsemltest/recommendation/test_ranking.py` to include tests for string `userCol` and `itemCol`.
  - Update `core/src/test/scala/com/microsoft/azure/synapse/ml/recommendation/SARSpec.scala` to include tests for string `userCol` and `itemCol`.

* **Documentation:**
  - Update `docs/Quick Examples/estimators/core/_Recommendation.md` to include examples with string `userCol` and `itemCol`.

---

For more details, open the [Copilot Workspace session](https://copilot-workspace.githubnext.com/microsoft/SynapseML/issues/2275?shareId=18fbc9a6-7678-44fa-ae21-613210d5c3f8).

acrolinxatmsft1 commented:
## Acrolinx Scorecards

**A minimum total score of 80 is required.**
 
Select the total score link to review all feedback on clarity, consistency, tone, brand, terms, spelling, grammar, readability, and inclusive language. _You should fix all spelling errors regardless of your total score_. Fixing spelling errors helps maintain customer trust in overall content quality.

| Article | Total score<br>(Required: 80) | Words + phrases<br>(Brand, terms) | Correctness<br>(Spelling, grammar) | Clarity<br>(Readability) |
|---------|:--------------:|:--------------------:|:------:|:---------:|
| [:white_check_mark:](# "Successfully processed") [docs/Quick Examples/estimators/core/_Recommendation.md](https://github.com/microsoft/SynapseML/blob/2bbf3cc8eee4ba05d30057d01fc685edf8b786bd/docs%2FQuick%20Examples%2Festimators%2Fcore%2F_Recommendation.md) | [72](https://microsoft-ce-csi.acrolinx.cloud/api/v1/checking/scorecards/e2557ee2-405b-4485-940f-4678d4c091f5) | 100 | 32 | 100 |

**More information about Acrolinx**

- [Install Acrolinx locally for VSCode](https://review.docs.microsoft.com/help/contribute/contribute-acrolinx-vscode)
- [False positives or issues](https://aka.ms/acrolinxbug)
- [Troubleshooting issues with Acrolinx](https://review.docs.microsoft.com/help/contribute/acrolinx-error-messages)

dciborow committed:
feat: add string to sar

acrolinxatmsft1 commented:
## Acrolinx Scorecards

**A minimum total score of 80 is required.**
 
Select the total score link to review all feedback on clarity, consistency, tone, brand, terms, spelling, grammar, readability, and inclusive language. _You should fix all spelling errors regardless of your total score_. Fixing spelling errors helps maintain customer trust in overall content quality.

| Article | Total score<br>(Required: 80) | Words + phrases<br>(Brand, terms) | Correctness<br>(Spelling, grammar) | Clarity<br>(Readability) |
|---------|:--------------:|:--------------------:|:------:|:---------:|
| [:white_check_mark:](# "Successfully processed") [docs/Quick Examples/estimators/core/_Recommendation.md](https://github.com/microsoft/SynapseML/blob/d58034492a2e8cc51c57d38d2019c93501c3023b/docs%2FQuick%20Examples%2Festimators%2Fcore%2F_Recommendation.md) | [72](https://microsoft-ce-csi.acrolinx.cloud/api/v1/checking/scorecards/0ce0abab-e4d3-4f7d-b4d8-cb6320a445bd) | 100 | 32 | 100 |

**More information about Acrolinx**

- [Install Acrolinx locally for VSCode](https://review.docs.microsoft.com/help/contribute/contribute-acrolinx-vscode)
- [False positives or issues](https://aka.ms/acrolinxbug)
- [Troubleshooting issues with Acrolinx](https://review.docs.microsoft.com/help/contribute/acrolinx-error-messages)

dciborow committed:
Delete core/src/main/python/synapse/ml/recommendation/SAR.py

acrolinxatmsft1 commented:
## Acrolinx Scorecards

**A minimum total score of 80 is required.**
 
Select the total score link to review all feedback on clarity, consistency, tone, brand, terms, spelling, grammar, readability, and inclusive language. _You should fix all spelling errors regardless of your total score_. Fixing spelling errors helps maintain customer trust in overall content quality.

| Article | Total score<br>(Required: 80) | Words + phrases<br>(Brand, terms) | Correctness<br>(Spelling, grammar) | Clarity<br>(Readability) |
|---------|:--------------:|:--------------------:|:------:|:---------:|
| [:white_check_mark:](# "Successfully processed") [docs/Quick Examples/estimators/core/_Recommendation.md](https://github.com/microsoft/SynapseML/blob/302fb7c689ff8b777b4668cf689191c4daad15ac/docs%2FQuick%20Examples%2Festimators%2Fcore%2F_Recommendation.md) | [72](https://microsoft-ce-csi.acrolinx.cloud/api/v1/checking/scorecards/37e7e2bc-4622-4cea-89a5-7743fa67d8fb) | 100 | 32 | 100 |

**More information about Acrolinx**

- [Install Acrolinx locally for VSCode](https://review.docs.microsoft.com/help/contribute/contribute-acrolinx-vscode)
- [False positives or issues](https://aka.ms/acrolinxbug)
- [Troubleshooting issues with Acrolinx](https://review.docs.microsoft.com/help/contribute/acrolinx-error-messages)

dciborow commented on core/src/main/scala/com/microsoft/azure/synapse/ml/recommendation/SAR.scala:
```diff
@@ -1,6 +1,3 @@
-// Copyright (C) Microsoft Corporation. All rights reserved.
-// Licensed under the MIT License. See LICENSE in project root for information.
-
 package com.microsoft.azure.synapse.ml.recommendation
```
```suggestion
// Copyright (C) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE in project root for information.

package com.microsoft.azure.synapse.ml.recommendation
```

dciborow commented on core/src/main/scala/com/microsoft/azure/synapse/ml/recommendation/SARModel.scala:
```diff
@@ -1,6 +1,3 @@
-// Copyright (C) Microsoft Corporation. All rights reserved.
-// Licensed under the MIT License. See LICENSE in project root for information.
-
 package com.microsoft.azure.synapse.ml.recommendation
```
```suggestion
// Copyright (C) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE in project root for information.

package com.microsoft.azure.synapse.ml.recommendation
```

dciborow committed:
Update core/src/main/scala/com/microsoft/azure/synapse/ml/recommendation/SARModel.scala

acrolinxatmsft1 commented:
## Acrolinx Scorecards

**A minimum total score of 80 is required.**
 
Select the total score link to review all feedback on clarity, consistency, tone, brand, terms, spelling, grammar, readability, and inclusive language. _You should fix all spelling errors regardless of your total score_. Fixing spelling errors helps maintain customer trust in overall content quality.

| Article | Total score<br>(Required: 80) | Words + phrases<br>(Brand, terms) | Correctness<br>(Spelling, grammar) | Clarity<br>(Readability) |
|---------|:--------------:|:--------------------:|:------:|:---------:|
| [:white_check_mark:](# "Successfully processed") [docs/Quick Examples/estimators/core/_Recommendation.md](https://github.com/microsoft/SynapseML/blob/8418cc06ac60f31dacc6865b7fd41356b74303eb/docs%2FQuick%20Examples%2Festimators%2Fcore%2F_Recommendation.md) | [72](https://microsoft-ce-csi.acrolinx.cloud/api/v1/checking/scorecards/70a923ec-8863-42e5-8741-79a76002fb2b) | 100 | 32 | 100 |

**More information about Acrolinx**

- [Install Acrolinx locally for VSCode](https://review.docs.microsoft.com/help/contribute/contribute-acrolinx-vscode)
- [False positives or issues](https://aka.ms/acrolinxbug)
- [Troubleshooting issues with Acrolinx](https://review.docs.microsoft.com/help/contribute/acrolinx-error-messages)

dciborow committed:
Update core/src/main/scala/com/microsoft/azure/synapse/ml/recommendation/SAR.scala

acrolinxatmsft1 commented:
## Acrolinx Scorecards

**A minimum total score of 80 is required.**
 
Select the total score link to review all feedback on clarity, consistency, tone, brand, terms, spelling, grammar, readability, and inclusive language. _You should fix all spelling errors regardless of your total score_. Fixing spelling errors helps maintain customer trust in overall content quality.

| Article | Total score<br>(Required: 80) | Words + phrases<br>(Brand, terms) | Correctness<br>(Spelling, grammar) | Clarity<br>(Readability) |
|---------|:--------------:|:--------------------:|:------:|:---------:|
| [:white_check_mark:](# "Successfully processed") [docs/Quick Examples/estimators/core/_Recommendation.md](https://github.com/microsoft/SynapseML/blob/f84e051e0ecfd429453715a1fbc6735485a9d4ba/docs%2FQuick%20Examples%2Festimators%2Fcore%2F_Recommendation.md) | [72](https://microsoft-ce-csi.acrolinx.cloud/api/v1/checking/scorecards/dde29b5c-1f69-4e8a-b78c-c5fde93699ab) | 100 | 32 | 100 |

**More information about Acrolinx**

- [Install Acrolinx locally for VSCode](https://review.docs.microsoft.com/help/contribute/contribute-acrolinx-vscode)
- [False positives or issues](https://aka.ms/acrolinxbug)
- [Troubleshooting issues with Acrolinx](https://review.docs.microsoft.com/help/contribute/acrolinx-error-messages)

dciborow commented:
/azp run

dciborow committed:
Update SARSpec.scala

acrolinxatmsft1 commented:
## Acrolinx Scorecards

**A minimum total score of 80 is required.**
 
Select the total score link to review all feedback on clarity, consistency, tone, brand, terms, spelling, grammar, readability, and inclusive language. _You should fix all spelling errors regardless of your total score_. Fixing spelling errors helps maintain customer trust in overall content quality.

| Article | Total score<br>(Required: 80) | Words + phrases<br>(Brand, terms) | Correctness<br>(Spelling, grammar) | Clarity<br>(Readability) |
|---------|:--------------:|:--------------------:|:------:|:---------:|
| [:white_check_mark:](# "Successfully processed") [docs/Quick Examples/estimators/core/_Recommendation.md](https://github.com/microsoft/SynapseML/blob/a38b134ca4a424f4175d3bc78f33348d7ec13333/docs%2FQuick%20Examples%2Festimators%2Fcore%2F_Recommendation.md) | [72](https://microsoft-ce-csi.acrolinx.cloud/api/v1/checking/scorecards/01bea800-ee66-462f-aa44-2870c70a305a) | 100 | 32 | 100 |

**More information about Acrolinx**

- [Install Acrolinx locally for VSCode](https://review.docs.microsoft.com/help/contribute/contribute-acrolinx-vscode)
- [False positives or issues](https://aka.ms/acrolinxbug)
- [Troubleshooting issues with Acrolinx](https://review.docs.microsoft.com/help/contribute/acrolinx-error-messages)

dciborow commented on core/src/test/scala/com/microsoft/azure/synapse/ml/recommendation/SARSpec.scala:
```diff
@@ -1,5 +1,5 @@
 // Copyright (C) Microsoft Corporation. All rights reserved.
-// Licensed under the MIT License. See LICENSE in project root for information.
+// Licensed under the MIT License. See LICENSE in the project root for information.
```
```suggestion
// Licensed under the MIT License. See LICENSE in project root for information.
```

dciborow committed:
Apply suggestions from code review

acrolinxatmsft1 commented:
## Acrolinx Scorecards

**A minimum total score of 80 is required.**
 
Select the total score link to review all feedback on clarity, consistency, tone, brand, terms, spelling, grammar, readability, and inclusive language. _You should fix all spelling errors regardless of your total score_. Fixing spelling errors helps maintain customer trust in overall content quality.

| Article | Total score<br>(Required: 80) | Words + phrases<br>(Brand, terms) | Correctness<br>(Spelling, grammar) | Clarity<br>(Readability) |
|---------|:--------------:|:--------------------:|:------:|:---------:|
| [:white_check_mark:](# "Successfully processed") [docs/Quick Examples/estimators/core/_Recommendation.md](https://github.com/microsoft/SynapseML/blob/129081da34f5c22eee8a2efef0456fa1cac412fe/docs%2FQuick%20Examples%2Festimators%2Fcore%2F_Recommendation.md) | [72](https://microsoft-ce-csi.acrolinx.cloud/api/v1/checking/scorecards/ff61d542-6a34-4564-9dfb-76dba0290e63) | 100 | 32 | 100 |

**More information about Acrolinx**

- [Install Acrolinx locally for VSCode](https://review.docs.microsoft.com/help/contribute/contribute-acrolinx-vscode)
- [False positives or issues](https://aka.ms/acrolinxbug)
- [Troubleshooting issues with Acrolinx](https://review.docs.microsoft.com/help/contribute/acrolinx-error-messages)

dciborow commented on docs/Quick Examples/estimators/core/_Recommendation.md:
```diff
@@ -61,6 +61,43 @@ ratings = (spark.createDataFrame([
     .dropDuplicates()
     .cache())
 
+ratings_with_strings = (spark.createDataFrame([
+      ("user0", "item1", 4, 4),
+      ("user0", "item3", 1, 1),
+      ("user0", "item4", 5, 5),
+      ("user0", "item5", 3, 3),
+      ("user0", "item7", 3, 3),
+      ("user0", "item9", 3, 3),
+      ("user0", "item10", 3, 3),
+      ("user1", "item1", 4, 4),
+      ("user1", "item2", 5, 5),
+      ("user1", "item3", 1, 1),
+      ("user1", "item6", 4, 4),
+      ("user1", "item7", 5, 5),
+      ("user1", "item8", 1, 1),
+      ("user1", "item10", 3, 3),
+      ("user2", "item1", 4, 4),
+      ("user2", "item2", 1, 1),
+      ("user2", "item3", 1, 1),
+      ("user2", "item4", 5, 5),
+      ("user2", "item5", 3, 3),
+      ("user2", "item6", 4, 4),
+      ("user2", "item8", 1, 1),
+      ("user2", "item9", 5, 5),
+      ("user2", "item10", 3, 3),
+      ("user3", "item2", 5, 5),
+      ("user3", "item3", 1, 1),
+      ("user3", "item4", 5, 5),
+      ("user3", "item5", 3, 3),
+      ("user3", "item6", 4, 4),
+      ("user3", "item7", 5, 5),
+      ("user3", "item8", 1, 1),
+      ("user3", "item9", 5, 5),
+      ("user3", "item10", 3, 3)
+      ], ["originalCustomerID", "newCategoryID", "rating", "notTime"])
+    .coalesce(1)
+    .cache())
+
```
```suggestion
```

dciborow commented on docs/Quick Examples/estimators/core/_Recommendation.md:
```diff
@@ -367,6 +482,10 @@ val adapter = (new RankingAdapter()
 val res1 = recommendationIndexer.fit(ratings).transform(ratings).cache()
 
 adapter.fit(res1).transform(res1).show()
+
+val res2 = recommendationIndexer.fit(ratings_with_strings).transform(ratings_with_strings).cache()
+
+adapter.fit(res2).transform(res2).show()
```
```suggestion
```

dciborow commented on docs/Quick Examples/estimators/core/_Recommendation.md:
```diff
@@ -344,6 +422,43 @@ val ratings = (Seq(
     .dropDuplicates()
     .cache())
 
+val ratings_with_strings = (Seq(
+      ("user0", "item1", 4, 4),
+      ("user0", "item3", 1, 1),
+      ("user0", "item4", 5, 5),
+      ("user0", "item5", 3, 3),
+      ("user0", "item7", 3, 3),
+      ("user0", "item9", 3, 3),
+      ("user0", "item10", 3, 3),
+      ("user1", "item1", 4, 4),
+      ("user1", "item2", 5, 5),
+      ("user1", "item3", 1, 1),
+      ("user1", "item6", 4, 4),
+      ("user1", "item7", 5, 5),
+      ("user1", "item8", 1, 1),
+      ("user1", "item10", 3, 3),
+      ("user2", "item1", 4, 4),
+      ("user2", "item2", 1, 1),
+      ("user2", "item3", 1, 1),
+      ("user2", "item4", 5, 5),
+      ("user2", "item5", 3, 3),
+      ("user2", "item6", 4, 4),
+      ("user2", "item8", 1, 1),
+      ("user2", "item9", 5, 5),
+      ("user2", "item10", 3, 3),
+      ("user3", "item2", 5, 5),
+      ("user3", "item3", 1, 1),
+      ("user3", "item4", 5, 5),
+      ("user3", "item5", 3, 3),
+      ("user3", "item6", 4, 4),
+      ("user3", "item7", 5, 5),
+      ("user3", "item8", 1, 1),
+      ("user3", "item9", 5, 5),
+      ("user3", "item10", 3, 3))
+    .toDF("originalCustomerID", "newCategoryID", "rating", "notTime")
+    .coalesce(1)
+    .cache())
+
```
```suggestion
```

dciborow commented on docs/Quick Examples/estimators/core/_Recommendation.md:
```diff
@@ -298,6 +372,10 @@ adapter = (RankingAdapter()
 res1 = recommendationIndexer.fit(ratings).transform(ratings).cache()
 
 adapter.fit(res1).transform(res1).show()
+
+res2 = recommendationIndexer.fit(ratings_with_strings).transform(ratings_with_strings).cache()
+
+adapter.fit(res2).transform(res2).show()
```
```suggestion
```

dciborow commented on docs/Quick Examples/estimators/core/_Recommendation.md:
```diff
@@ -275,6 +312,43 @@ ratings = (spark.createDataFrame([
     .dropDuplicates()
     .cache())
 
+ratings_with_strings = (spark.createDataFrame([
+      ("user0", "item1", 4, 4),
+      ("user0", "item3", 1, 1),
+      ("user0", "item4", 5, 5),
+      ("user0", "item5", 3, 3),
+      ("user0", "item7", 3, 3),
+      ("user0", "item9", 3, 3),
+      ("user0", "item10", 3, 3),
+      ("user1", "item1", 4, 4),
+      ("user1", "item2", 5, 5),
+      ("user1", "item3", 1, 1),
+      ("user1", "item6", 4, 4),
+      ("user1", "item7", 5, 5),
+      ("user1", "item8", 1, 1),
+      ("user1", "item10", 3, 3),
+      ("user2", "item1", 4, 4),
+      ("user2", "item2", 1, 1),
+      ("user2", "item3", 1, 1),
+      ("user2", "item4", 5, 5),
+      ("user2", "item5", 3, 3),
+      ("user2", "item6", 4, 4),
+      ("user2", "item8", 1, 1),
+      ("user2", "item9", 5, 5),
+      ("user2", "item10", 3, 3),
+      ("user3", "item2", 5, 5),
+      ("user3", "item3", 1, 1),
+      ("user3", "item4", 5, 5),
+      ("user3", "item5", 3, 3),
+      ("user3", "item6", 4, 4),
+      ("user3", "item7", 5, 5),
+      ("user3", "item8", 1, 1),
+      ("user3", "item9", 5, 5),
+      ("user3", "item10", 3, 3)
+      ], ["originalCustomerID", "newCategoryID", "rating", "notTime"])
+    .coalesce(1)
+    .cache())
+
```
```suggestion
```

dciborow committed:
Apply suggestions from code review

acrolinxatmsft1 commented:
## Acrolinx Scorecards

**A minimum total score of 80 is required.**
 
Select the total score link to review all feedback on clarity, consistency, tone, brand, terms, spelling, grammar, readability, and inclusive language. _You should fix all spelling errors regardless of your total score_. Fixing spelling errors helps maintain customer trust in overall content quality.

| Article | Total score<br>(Required: 80) | Words + phrases<br>(Brand, terms) | Correctness<br>(Spelling, grammar) | Clarity<br>(Readability) |
|---------|:--------------:|:--------------------:|:------:|:---------:|
| [:white_check_mark:](# "Successfully processed") [docs/Quick Examples/estimators/core/_Recommendation.md](https://github.com/microsoft/SynapseML/blob/fc8a97fa5d381e52c0a65c815d83da6c867981aa/docs%2FQuick%20Examples%2Festimators%2Fcore%2F_Recommendation.md) | [72](https://microsoft-ce-csi.acrolinx.cloud/api/v1/checking/scorecards/f97c1074-31da-4038-93e8-7771e1ded382) | 100 | 32 | 100 |

**More information about Acrolinx**

- [Install Acrolinx locally for VSCode](https://review.docs.microsoft.com/help/contribute/contribute-acrolinx-vscode)
- [False positives or issues](https://aka.ms/acrolinxbug)
- [Troubleshooting issues with Acrolinx](https://review.docs.microsoft.com/help/contribute/acrolinx-error-messages)

dciborow committed:
Update docs/Quick Examples/estimators/core/_Recommendation.md

dciborow commented on core/src/main/python/synapse/ml/recommendation/SARModel.py:
```diff
@@ -15,3 +15,8 @@
 class SARModel(_SARModel):
     def recommendForAllUsers(self, numItems):
         return self._call_java("recommendForAllUsers", numItems)
+
+    def recommendForUserSubset(self, dataset, numItems):
+        if dataset.schema[self.getUserCol()].dataType == StringType():
+            dataset = dataset.withColumn(self.getUserCol(), dataset[self.getUserCol()].cast("int"))
+        return self._call_java("recommendForUserSubset", dataset, numItems)
```
```suggestion
```

dciborow committed:
Apply suggestions from code review

dciborow committed:
Update SAR.scala to handle integer types for userId and itemId

* **SAR.scala**
  - Update `calculateUserItemAffinities` method to handle integer types for `userId` and `itemId`
  - Update `calculateItemItemSimilarity` method to handle integer types for `userId` and `itemId`

* **test_ranking.py**
  - Add test cases to verify the functionality of SAR model with integer types for `userId` and `itemId`

dciborow commented:
/azp run

dciborow committed:
Discard changes to core/src/test/scala/com/microsoft/azure/synapse/ml/recommendation/SARSpec.scala

dciborow committed:
Add new test cases for User Column with Strings and other datatypes in SARSpec.scala

* Add a test case for handling User Column with Strings
* Add a test case for handling User Column with different datatypes
* Verify the handling of User Column with Strings and other datatypes in SAR.scala
* Ensure the new test cases are concise and focused on the new code
* Place the new test cases in an appropriate location within the file
